backend/src/datasource/user-datasource.js

```diff
@@ -1,46 +1,101 @@
+    ];
+  }
 
-    constructor(users = null) {
-        super()
-
-        this.users = users || [
-            {
-                id: 1,
-                name: 'Robert',
-            },
-            {
-                id: 2,
-                name: 'Youri',
-            }
-        ]
-    }
+  getUserById(id) {
+    return this.users.find((user) => user.id == id);
+  }
 
-    getUserById(id) {
-        return this.users.find(user => user.id == id);
-    }
+  getUserByName(name) {
+    return this.users.find((user) => user.name == name);
+  }
 
-    getUserByName(name) {
-        return this.users.find(user => user.name == name);
-    }
+  getUserByEmail(email) {
+    return this.users.find(
+      (user) => user.email.toLowerCase() === email.toLowerCase()
+    );
+  }
 
-    allUsers() {
-        return this.users;
-    }
+  allUsers() {
+    return this.users;
+  }
+
+  userExists(userId) {
+      let user = this.getUserById(userId)
+      return user ? true : false
+  }
 
-    addUser(name) {
-        if (this.getUserByName(name) == undefined) {
-            this.users.push({
-                id : Math.max(...this.users.map(user => user.id), 0) + 1,
-                name: name
-            });
-        }
+  addUser(name, email, password) {
+    this.users.push(
+      new User(
+        Math.max(...this.users.map((user) => user.id), 0) + 1,
+        name,
+        email,
+        password
+      )
+    );
+  }
+
+  removeUser(id) {
+    this.users = this.users.filter((user) => {
+      return user.id != id;
+    });
+  }
+
+  async signup(name, email, password) {
+    if (this.getUserByEmail(email) == undefined) {
+      if (this.passwordValid(password)) {
+        let jwtToken = await bcrypt
+          .genSalt(10)
+          .then((salt) => {
+            return bcrypt.hash(password, salt);
+          })
+          .then((hash) => {
+            this.addUser(name, email, hash)
+            return this.createJWT(this.getUserByEmail(email))
+          });
+        
+        return jwtToken;
+      }
+      return "Password invalid. Must be at least 8 characters long. User not added.";
     }
+    return "Email already exists. User not added.";
+  }
+
+  passwordValid(password) {
+    return password.length >= 8;
+  }
 
-    removeUser(id) {
-        this.users = this.users.filter((user) => {
-            return user.id != id;
-        });
+  login(email, password) {
+    let user = this.getUserByEmail(email);
+
+    if (user) {
+        let result = bcrypt.compare(password, user.password)
+            .then(res => {
+                return res ? this.createJWT(user) : "User or Password incorrect"
+            }).catch(err => {
+                console.log(err.message)
+                return "User or Password incorrect"
+            })
+        return result
     }
-}
\ No newline at end of file
+
+    return "User or Password incorrect"
+  }
+
+  createJWT(user) {
```

Comment:
* Do not expose password hashes in the JWT payload. Also do not expose private information like `email` unnecessarily.